Allow multiple depth

psealock opened the pull request:
This PR makes changes to the PHP registration such that `add_category` is an actual category and not necessarily a top level category.

### `add_category` 

`\Automattic\WooCommerce\Navigation\Menu::add_category` now always maps to a `<NavigationItem />` without a url property so the chevron is rendered. Its location is determined by the `parent` property.

![Screen Shot 2020-09-29 at 8 58 38 PM](https://user-images.githubusercontent.com/1922453/94529739-969aa800-0296-11eb-8f47-309471e75695.png)

At the same time, its `id` is used to create a `<NavigationMenu />` which houses any child items that may be registered.

![Screen Shot 2020-09-29 at 9 00 52 PM](https://user-images.githubusercontent.com/1922453/94529969-e4afab80-0296-11eb-9b9d-95869bce5c4f.png)

### `add_item` 

`\Automattic\WooCommerce\Navigation\Menu::add_item` always maps to a `<NavigationItem />`. Its location is also determined by the `parent` property.

### Previously

`add_category` was used to signify a top level item which could be a singular item or a parent item that navigates to a new menu, such as "Analytics". The motivation here was to discourage use of `add_category` so that plugins wouldn't abuse the top level navigation to house their links.

I find this duality difficult to reason about and a more direct representation to be more effective. But, there are some changes we can make to make it even more clear:

### Proposal

1. Change `add_category` to `add_menu` to clarify the distinction that a menu is used to create a navigational item that has children.

2. `add_category` defaults to a parent of `woocommerce` but should be an empty string. I left this as is because otherwise changes would be far reaching and involve wc-admin. I'd like to have buying before making those changes.

3. Find another way to discourage addition of top level menus. How might we do this? Or are we kidding ourselves in believing we have the power to limit abuse? I'm open to suggestions here.

### Test

Pull the latest from https://github.com/psealock/examples, which has added a multi-level navigation.

psealock commented:
>Find another way to discourage addition of top level menus. How might we do this?

What if we created another method reserved for top level items? Or passed a parameter which must be true to designate top level?

```php
Menu::add_item( array( /* config here */ ), true );  // top level item
Menu::add_item( array( /* config here */ ) );  // regular item
```

joshuatf commented:
>Change add_category to add_menu to clarify the distinction that a menu is used to create a navigational item that has children.

I think `add_menu` sounds like a whole new menu is being added as opposed to a level or "connected menu" (not sure on the current semantics in the GB nav component), but this could be confusing for consumers.  Keeping it as the `add_category` you've updated to sounds good to me.

>add_category defaults to a parent of woocommerce but should be an empty string. I left this as is because otherwise changes would be far reaching and involve wc-admin. I'd like to have buying before making those changes.

The idea here is that changing it to an empty string would force consumers to specify the parent, right?

>Find another way to discourage addition of top level menus. How might we do this? Or are we kidding ourselves in believing we have the power to limit abuse? I'm open to suggestions here.

We may be kidding ourselves that this is going to prevent abuse, but trying does little harm if the semantics aren't overly confusing.  

>Menu::add_item( array( /* config here */ ), true );

One potential problem I see with this is that users may copy and paste code from here, not noticing the second argument, and call it a day.  I think the route we use should be explicit so that users must acknowledge what they're doing.  Hopefully this prevents a marginal amount of abuse.

In order to do this, I think we must define the intent of the top-level items.  Originally, I think we were referring to these as "buckets."  If we can convey through a parameter that this area is meant for this type of item, maybe it will discourage users from adding any items they please here.

`Menu::add_item( array( 'is_top_level_group' => true ) );`
* `is_top_level_group`
* `is_top_level_bucket`
* `is_top_level_section`

We could even alias this name to a function and use those to make the intent of these items even more transparent.

```
function add_top_level_group( args ) {
  self::add_item( array_merge( args, 'top_level' => true );
}
```

/cc @jameskoster in case you have any wisdom to share on the semantics of the buckets/groups.

psealock commented:
>The idea here is that changing it to an empty string would force consumers to specify the parent, right?

Right, defaulting to an empty string forces consumers to specify the parent and deters piling items onto the top level by default.

>If we can convey through a parameter that this area is meant for this type of item, maybe it will discourage users from adding any items they please here.

I think `is_top_level_item` works since item is the thing being added. `Menu::add_category` could have the same treatment.

```
Menu::add_item( array( 'is_top_level_item' => true ) );
```

psealock commented:
## Update

I have kept the methods the same, but added options properties `is_top_level_category` and `is_top_level_item` for adding categories and items, respectively. These must be set to true in order to have `woocommerce` as a parent. Otherwise, the default will be used.

Tests are also updated.

I also removed wc-admin specific code from the container component.

### Test

You can test against the `update/nav-item-registration` branch in wc-admin (PR to come) and the lasted `master` branch from https://github.com/psealock/examples to see multi-level at play.

joshuatf commented on src/CoreMenu.php:
```diff
@@ -78,23 +78,25 @@ public function add_core_items() {
+				'migrate'               => false,
+				'menuId'                => 'secondary',
+				'order'                 => 20,
+				'is_top_level_category' => true,
```
Small optional nitpick, I think we can update this to be `is_top_level` so this is consistent across items and categories.

psealock replied:
Good idea. Changed to `is_top_level` in https://github.com/woocommerce/navigation/pull/117/commits/1be2ded157ab36f111c57b69830ef9d1c3a63a5e

joshuatf reviewed (approved):
Pre-approving pending one optional change for the property names.  Thanks for the changes, Paul!
